File: src/terceraParteEjerciciosTraduccion/ejercicio3/ejercicio3.java
package terceraParteEjerciciosTraduccion.ejercicio3;

//3. Juego de Rol (RPG)
// Crea un sistema básico para un juego de rol. Define clases para Personaje,
// Guerrero, Mago, y Hechizo. Implementa métodos para atacar y usar hechizos. Los
// personajes deben tener puntos de vida y poder realizar acciones según su clase.
import java.util.Scanner;

abstract class Personaje {
    String nombre;
    int puntosVida;

    public Personaje(String nombre, int puntosVida) {
        this.nombre = nombre;
        this.puntosVida = puntosVida;
    }

    public abstract void atacar(Personaje enemigo);

    public boolean estaVivo() {
        return puntosVida > 0;
    }
}

class Guerrero extends Personaje {
    int fuerza;

    public Guerrero(String nombre, int puntosVida, int fuerza) {
        super(nombre, puntosVida);
        this.fuerza = fuerza;
    }

    @Override
    public void atacar(Personaje enemigo) {
        System.out.println(nombre + " ataca con fuerza de " + fuerza);
        enemigo.puntosVida -= fuerza;
        System.out.println("El enemigo ahora tiene " + enemigo.puntosVida + " puntos de vida.");
    }
}

class Mago extends Personaje {
    int poderMagico;

    public Mago(String nombre, int puntosVida, int poderMagico) {
        super(nombre, puntosVida);
        this.poderMagico = poderMagico;
    }

    public void lanzarHechizo(Personaje enemigo) {
        System.out.println(nombre + " lanza un hechizo con poder de " + poderMagico);
        enemigo.puntosVida -= poderMagico;
        System.out.println("El enemigo ahora tiene " + enemigo.puntosVida + " puntos de vida.");
    }

    @Override
    public void atacar(Personaje enemigo) {
        lanzarHechizo(enemigo);
    }
}

class JuegoRPG {
    public static void main(String[] args) {
        Scanner scanner = new Scanner(System.in);

        Guerrero guerrero = new Guerrero("Guerrero", 100, 15);
        Mago mago = new Mago("Mago", 80, 20);

        Personaje enemigo = new Guerrero("Enemigo", 60, 10);

        int opcion;
        do {
            System.out.println("\n1. Atacar como Guerrero");
            System.out.println("2. Lanzar Hechizo como Mago");
            System.out.println("3. Salir");
            System.out.print("Elige una opción: ");
            opcion = scanner.nextInt();

            switch (opcion) {
                case 1:
                    guerrero.atacar(enemigo);
                    break;
                case 2:
                    mago.lanzarHechizo(enemigo);
                    break;
                case 3:
                    System.out.println("Saliendo del juego...");
                    break;
                default:
                    System.out.println("Opción no válida.");
            }

            if (!enemigo.estaVivo()) {
                System.out.println("¡Has derrotado al enemigo!");
                break;
            }

        } while (opcion != 3);

        scanner.close();
    }
}

File: src/terceraParteEjerciciosTraduccion/ejercicio3/ejercicio3.kt
package terceraParteEjerciciosTraduccion.ejercicio3

import kotlin.system.exitProcess

abstract class Personaje2(val nombre: String, var puntosVida: Int) {
    abstract fun atacar(enemigo: Personaje2)
    fun estaVivo() = puntosVida > 0
}

class Guerrero2(nombre: String, puntosVida: Int, private val fuerza: Int) : Personaje2(nombre, puntosVida) {
    override fun atacar(enemigo: Personaje2) {
        println("$nombre ataca con fuerza de $fuerza")
        enemigo.puntosVida -= fuerza
        println("El enemigo ahora tiene ${enemigo.puntosVida} puntos de vida.")
    }
}

class Mago2(nombre: String, puntosVida: Int, private val poderMagico: Int) : Personaje2(nombre, puntosVida) {
    fun lanzarHechizo(enemigo: Personaje2) {
        println("$nombre lanza un hechizo con poder de $poderMagico")
        enemigo.puntosVida -= poderMagico
        println("El enemigo ahora tiene ${enemigo.puntosVida} puntos de vida.")
    }

    override fun atacar(enemigo: Personaje2) {
        lanzarHechizo(enemigo)
    }
}

fun main() {
    val guerrero = Guerrero2("Guerrero", 100, 15)
    val mago = Mago2("Mago", 80, 20)
    val enemigo = Guerrero2("Enemigo", 60, 10)

    while (true) {
        println("\n1. Atacar como Guerrero")
        println("2. Lanzar Hechizo como Mago")
        println("3. Salir")
        print("Elige una opción: ")
        when (readlnOrNull()?.toIntOrNull()) {
            1 -> guerrero.atacar(enemigo)
            2 -> mago.lanzarHechizo(enemigo)
            3 -> {
                println("Saliendo del juego...")
                exitProcess(0)
            }
            else -> println("Opción no válida.")
        }

        if (!enemigo.estaVivo()) {
            println("¡Has derrotado al enemigo!")
            break
        }
    }
}
